examples/step-86/step-86.cc

```diff
@@ -0,0 +1,896 @@
+      {
+        static std::vector<std::pair<double, std::string>> times_and_names;
+        times_and_names.emplace_back(time, filename);
+        std::ofstream pvd_output("solution.pvd");
+        DataOutBase::write_pvd_record(pvd_output, times_and_names);
+      }
+  }
+
+
+  // This function is essentially identical to step-26, with the only difference
+  // that it uses the parallel::distributed::GridRefinement namespace function
+  // instead of the serial one. Once again, we make sure that we never fall
+  // below the minimum refinement level, and above the maximum one, that we can
+  // select from the parameter file.
+  template <int dim>
+  void HeatEquation<dim>::prepare_for_coarsening_and_refinement(
+    const PETScWrappers::MPI::Vector &sol)
+  {
+    Vector<float> estimated_error_per_cell(triangulation.n_active_cells());
+
+    locally_relevant_solution = sol;
+    KellyErrorEstimator<dim>::estimate(dof_handler,
+                                       QGauss<dim - 1>(fe.degree + 1),
+                                       {},
+                                       locally_relevant_solution,
+                                       estimated_error_per_cell);
+
+    parallel::distributed::GridRefinement::refine_and_coarsen_fixed_fraction(
+      triangulation, estimated_error_per_cell, 0.6, 0.4);
+
+    const auto max_grid_level =
+      initial_global_refinement + max_delta_refinement_level;
+    const auto min_grid_level = initial_global_refinement;
+
+    if (triangulation.n_levels() > max_grid_level)
+      for (const auto &cell :
+           triangulation.active_cell_iterators_on_level(max_grid_level))
+        cell->clear_refine_flag();
+    for (const auto &cell :
+         triangulation.active_cell_iterators_on_level(min_grid_level))
+      cell->clear_coarsen_flag();
+  }
+
+
+  // This function is called by the time stepper whenever it requires to
+  // transfer the solution and any intermediate stage vectors to a new mesh. We
+  // must make sure that all input vectors are transformed into ghosted vectors
+  // before the actual transfer is executed, and that we distribute the hanging
+  // node constraints on the output vectors as soon as we have interpolated the
+  // vectors to the new mesh.
+  //
+  // We have no way to enforce boundary conditions at this stage, since every
+  // type of time advancing scheme may have computed the intermediate stage at
+  // different times, and we have no way to know what boundary conditions to use
+  // at what stage.
+  //
+  // While this could be a problem if we used the values of the solution and of
+  // the intermediate stages on the constrained degrees of freedom to compute
+  // the errors, we do not do so. Instead, we compute the errors on the
+  // differential equation, and ignore algebraic constraints, therefore we do no
+  // need to guarantee that the boundary conditions are satisfied also in the
+  // intermediate stages.
+  //
```

Comment:
```suggestion
```

Reply:
I don't think this is correct. The issue we were facing was different, and probably related to the fact that the adaptor was complaining about new dofs on the boundary (that we now exclude from the local truncation error computation)